### ISSUE TITLE
False positive explicit_into_iter_loop on an into_iter method with type parameter

### ISSUE THREAD

**Issue by dtolnay** [rust-lang/rust-clippy#6900]
As of most recent nightly:

```rust
#![deny(clippy::explicit_into_iter_loop)]

use std::marker::PhantomData;

pub struct S;

impl S {
    #[allow(clippy::should_implement_trait)]
    pub fn into_iter<T>(self) -> I<T> {
        let _ = self;
        I { _marker: PhantomData }
    }
}

pub struct I<T> {
    _marker: PhantomData<T>,
}

impl<T> Iterator for I<T> {
    type Item = T;
    fn next(&mut self) -> Option<Self::Item> {
        None
    }
}

fn main() {
    for t in S.into_iter::<i32>() {
        println!("{}", t);
    }
}
```

```console
error: it is more concise to loop over containers instead of using explicit iteration methods
  --> src/main.rs:27:14
   |
27 |     for t in S.into_iter::<i32>() {
   |              ^^^^^^^^^^^^^^^^^^^^ help: to write this more concisely, try: `S`
   |
note: the lint level is defined here
  --> src/main.rs:1:9
   |
1  | #![deny(clippy::explicit_into_iter_loop)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#explicit_into_iter_loop
```

Clippy wants to see this rewritten as:

```diff
-     for t in S.into_iter::<i32>() {
+     for t in S {
```

However, that replacement code does not compile. Therefore I believe the `clippy::explicit_into_iter_loop` lint should be fixed to not trigger on this code.

```console
error[E0277]: `S` is not an iterator
  --> src/main.rs:27:14
   |
27 |     for t in S {
   |              ^ `S` is not an iterator
   |
   = help: the trait `std::iter::Iterator` is not implemented for `S`
   = note: required because of the requirements on the impl of `std::iter::IntoIterator` for `S`
   = note: required by `std::iter::IntoIterator::into_iter`
```

<br>

### Meta

- `cargo clippy -V`: e.g. clippy 0.1.52 (acca818 2021-03-13)
- `rustc -Vv`:
  ```
  rustc 1.52.0-nightly (acca81892 2021-03-13)
  binary: rustc
  commit-hash: acca818928654807ed3bc1ce0e97df118f8716c8
  commit-date: 2021-03-13
  host: x86_64-unknown-linux-gnu
  release: 1.52.0-nightly
  LLVM version: 12.0.0
  ```
